Exception when calling toListL and similar functions on response streams without mapping them first

This might be an issue with the Monix Observable library but I'd like to verify on our end first.

Given the following program:
```
import cats.effect._
import freestyle.rpc._
import freestyle.rpc.protocol._
import monix.execution.Scheduler
import monix.reactive.Observable

@message case class Bar(id: Int)

@service trait Foo[F[_]] {
  @rpc(Protobuf) @stream[ResponseStreaming.type]
  def bars(request: Empty.type): Observable[Bar]
}

class FooHandler[F[_]: Async](implicit S: Scheduler) extends Foo[F] {
  def bars(request: Empty.type): Observable[Bar] = Observable.fromIterable(Seq(1, 2, 3).map(Bar))
}

object Server {
  import freestyle.rpc.server._
  import Scheduler.Implicits.global

  val port: Int = defaultPort
  implicit val serverApp: ServerW = {
    implicit val serviceHandler = new FooHandler[IO]
    ServerW.default(port, List(AddService(Foo.bindService[IO])))
  }

  def main(args: Array[String]): Unit = {
    import freestyle.rpc.server.implicits._
    server[IO].unsafeRunSync
  }
}

object Client {
  import freestyle.rpc.client.implicits._
  import Scheduler.Implicits.global

  val client: Foo.Client[IO] = Foo.client[IO](ChannelForAddress("localhost", Server.port))

  def main(args: Array[String]): Unit = {
    val stream: Observable[Bar] = client.bars(Empty)
    val elements: List[Bar] = stream.toListL.to[IO].unsafeRunSync
    println(elements)
  }
}

```
the client will crash with the following exception:
```
Exception in thread "main" io.grpc.StatusRuntimeException: CANCELLED: Failed to read message.
	at io.grpc.Status.asRuntimeException(Status.java:526)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:419)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41)
	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:684)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:41)
	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:391)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:475)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:63)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:557)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$600(ClientCallImpl.java:478)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:542)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException: onSubscription never happened, see rule 2.13 in the Reactive Streams spec
	at monix.reactive.internal.rstreams.SyncSubscriberAsReactiveSubscriber.onNext(SubscriberAsReactiveSubscriber.scala:251)
	at monix.reactive.internal.builders.ReactiveObservable$$anon$1.onNext(ReactiveObservable.scala:39)
	at freestyle.rpc.internal.MonixAdapters$$anon$3$$anon$6.onNext(MonixAdapters.scala:55)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onMessage(ClientCalls.java:406)
	at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:36)
	at io.grpc.ForwardingClientCallListener.onMessage(ForwardingClientCallListener.java:36)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1MessagesAvailable.runInContext(ClientCallImpl.java:530)
	... 5 more
```

However, changing the line:
`val elements: List[Bar] = stream.toListL.to[IO].unsafeRunSync`
to:
`val elements: List[Bar] = stream.map(identity).toListL.to[IO].unsafeRunSync`
will make it run without problems.

Internally, this occurs whenever the initial transformation creates a subscriber that subclasses `Subscriber.Sync`, which causes `SubscriberAsReactiveSubscriber.apply()` to create a `SyncSubscriberAsReactiveSubscriber`, for which the publisher seems to fail to call `onSubscribe()`, causing a crash on `onNext()`. Calls which generate an `AsyncSubscriberAsReactiveSubscriber` don't have this problem.

The issue doesn't occur with non-RPC tests such as:
```
    val stream: Observable[Int] = Observable.fromIterable(Seq(1, 2, 3))
    val elements: List[Int] = stream.toListL.to[IO].unsafeRunSync
```

This affects some `Observable`->`Task` transformations such as `toListL`, `firstL`, or `completedL`, but not others such as `isEmptyL`. As seen above, the problem goes away when calling any `Observable`->`Observable` transformation beforehand, even as trivial as `map(identity)`.

--- a/modules/internal/src/main/scala/client/package.scala
+++ b/modules/internal/src/main/scala/client/package.scala
@@ -19,6 +19,7 @@ package internal
 
 import io.grpc.{CallOptions, Channel, MethodDescriptor}
 import io.grpc.stub.ClientCalls
+import monix.execution.rstreams.Subscription
 import org.reactivestreams.{Publisher, Subscriber}
 
 package object client {
@@ -33,6 +34,7 @@ package object client {
     new Publisher[Res] {
       override def subscribe(s: Subscriber[_ >: Res]): Unit = {
         val subscriber: Subscriber[Res] = s.asInstanceOf[Subscriber[Res]]
+        s.onSubscribe(Subscription.empty)
         ClientCalls.asyncServerStreamingCall(
           channel.newCall[Req, Res](descriptor, options),
           request,

--- a/modules/server/src/test/scala/protocol/RPCTests.scala
+++ b/modules/server/src/test/scala/protocol/RPCTests.scala
@@ -467,6 +467,15 @@ class RPCTests extends RpcBaseTestSuite with BeforeAndAfterAll {
 
     }
 
+    "#192 issue - be able to run server streaming services using .toListL without mapping" in {
+
+      def clientProgram[F[_]](implicit APP: MyRPCClient[F]): F[List[C]] =
+        APP.ss192(a2.x, a2.y)
+
+      clientProgram[ConcurrentMonad].unsafeRunSync() shouldBe cList
+
+    }
+
   }
 
 }

--- a/modules/server/src/test/scala/protocol/Utils.scala
+++ b/modules/server/src/test/scala/protocol/Utils.scala
@@ -167,6 +167,7 @@ object Utils extends CommonUtils {
       def uws(x: Int, y: Int): F[C]
       def uwe(a: A, err: String): F[C]
       def ss(a: Int, b: Int): F[List[C]]
+      def ss192(a: Int, b: Int): F[List[C]]
       def sswe(a: A, err: String): F[List[C]]
       def cs(cList: List[C], bar: Int): F[D]
       def bs(eList: List[E]): F[E]
@@ -383,6 +384,12 @@ object Utils extends CommonUtils {
             .toListL
             .to[F]
 
+        override def ss192(a: Int, b: Int): F[List[C]] =
+          client
+            .serverStreaming(B(A(a, a), A(b, b)))
+            .toListL
+            .to[F]
+
         override def sswe(a: A, err: String): F[List[C]] =
           client
             .serverStreamingWithError(E(a, err))
@@ -489,6 +496,12 @@ object Utils extends CommonUtils {
                 debug(s"[CLIENT] Result #$i: $c")
                 c
             }
+            .toListL
+            .to[F]
+
+        override def ss192(a: Int, b: Int): F[List[C]] =
+          client
+            .serverStreamingCompressed(B(A(a, a), A(b, b)))
             .toListL
             .to[F]
 

